discovery-provider/src/solana/anchor_program_indexer.py

```diff
@@ -61,23 +71,299 @@ def get_latest_slot(self):
         return latest_slot
 
     def validate_and_save_parsed_tx_records(
-        self, processed_transactions, metadata_dictionary
+        self, processed_transactions: List[Any], metadata_dictionary: Dict
     ):
         self.msg(
             f"validate_and_save anchor {processed_transactions} - {metadata_dictionary}"
         )
-        # TODO: Conditionally add database modifications here depending on transaction information
+        self.msg(f"{processed_transactions}")
         with self._db.scoped_session() as session:
-            for transaction in processed_transactions:
-                session.add(
+            session.bulk_save_objects(
+                [
                     AudiusDataTx(
                         signature=transaction["tx_sig"],
                         slot=transaction["result"]["slot"],
                     )
+                    for transaction in processed_transactions
+                ]
+            )
+
+            # Find user ids in DB and create dictionary mapping
+            db_models: Dict = defaultdict(lambda: {})
+            if metadata_dictionary["ids"]["users"]:
+                existing_users = (
+                    session.query(User)
+                    .filter(
+                        User.is_current,
+                        User.user_id.in_(metadata_dictionary["ids"]["users"]),
+                    )
+                    .all()
                 )
+                for user in existing_users:
+                    db_models["users"][user.user_id] = user
+
+            self.process_transactions(
+                session, processed_transactions, db_models, metadata_dictionary
+            )
+
+    def process_transactions(
+        self,
+        session: Session,
+        processed_transactions: List[Any],
+        db_models: Dict,
+        metadata_dictionary: Dict,
+    ):
+        id_dict: Dict[str, set] = defaultdict(lambda: set())
+        records = []
+        for transaction in processed_transactions:
+            instructions = transaction.get("tx_metadata").get("instructions")
+            for instruction in instructions:
+                instruction_name = instruction.get("instruction_name")
```

Comment:
Discussing ways to break this up. Maybe define an Instruction base class and implement parsing / db updates / validation for each instruction type.